### PR TITLE
[WIP] add Identity and Access Management service

### PR DESCRIPTION
**Pull request by chai2010**
update #558

### CODE REVIEW

File: api/iam/account.proto [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+message GroupMember {
+	google.protobuf.StringValue group_id = 1;
+	google.protobuf.StringValue user_id = 2;
+}
+
+service AccountManager {
+	rpc DescribeUsers(DescribeUsersRequest) returns (DescribeUsersResponse);
+	rpc ModifyUser(ModifyUserRequest) returns (ModifyUserResponse);
+	rpc DeleteUsers(DeleteUsersRequest) returns (DeleteUsersResponse);
+	rpc InviteUsers(InviteUsersRequest) returns (InviteUsersResponse);
```
Comment (chai2010) [inline api/iam/account.proto]: remove InviteUsers

File: api/iam/account.proto [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+message InviteUsersResponse {
+	repeated string token = 1;
+}
+message CreatePasswordResetRequest {
+	google.protobuf.StringValue user_id = 1;
+	google.protobuf.StringValue password = 2;
+}
+message CreatePasswordResetResponse {
+	google.protobuf.StringValue user_id = 1;
+	google.protobuf.StringValue reset_id = 2;
```
Comment (chai2010) [inline api/iam/account.proto]: reset_id is not jwt token

File: api/iam/account.proto [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+}
+message ChangePasswordRequest {
+	google.protobuf.StringValue new_password = 1;
+	google.protobuf.StringValue reset_id = 2;
+}
+message ChangePasswordResponse {
+	google.protobuf.StringValue user_id = 1;
+}
+message GetPasswordResetRequest {
+	google.protobuf.StringValue reset_id = 1;
```
Comment (chai2010) [inline api/iam/account.proto]: for supper user, user info is grpc's context.

TODO

File: api/iam/account.proto [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+}
+message GetPasswordResetResponse {
+	google.protobuf.StringValue reset_id = 1;
+	google.protobuf.StringValue user_id = 2;
+}
+message CreateUserRequest {
+	google.protobuf.StringValue email = 1;
```
Comment (chai2010) [inline api/iam/account.proto]: username is prefix(email)

File: api/iam/account.proto [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	google.protobuf.StringValue password = 2;
+	google.protobuf.StringValue role = 3; // default is user
+}
+message CreateUserResponse {
+	google.protobuf.StringValue user_id = 1;
+}
+
+message ValidateUserPasswordRequest {
+	string email = 1;
```
Comment (chai2010) [inline api/iam/account.proto]: assert(username == prefix(email))

File: api/iam/account.proto [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	string password = 2;
+}
+message ValidateUserPasswordResponse {
+	bool validated = 1;
+}
+
+message CreateGroupRequest {
+	google.protobuf.StringValue name = 1;
```
Comment (chai2010) [inline api/iam/account.proto]: add desc field

File: api/iam/account.proto [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+message DescribeGroupsRequest {
+	google.protobuf.StringValue search_word = 1;
+	google.protobuf.StringValue sort_key = 2;
+	google.protobuf.BoolValue reverse = 3;
+	repeated string group_id = 4;
+	repeated string user_id = 5;
+	uint32 limit = 6;
+	uint32 offset = 7;
```
Comment (chai2010) [inline api/iam/account.proto]: 1-10 for common fields, limit/offset are common fields

File: pkg/config/config.go [hunk 8]
```diff
@@ -33,6 +34,11 @@ type EtcdConfig struct {
 	Endpoints string `default:"openpitrix-etcd:2379"` // Example: "localhost:2379,localhost:22379,localhost:32379"
 }
 
+type TokenConfig struct {
+	Secret          string `default:"token_secret"`
+	DurationSeconds int    `default:"3600"`
```
Comment (chai2010) [inline pkg/config/config.go]: s/int/time.Duration

File: pkg/db/schema/iam/V0_1__init.sql [hunk 9]
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE IF NOT EXISTS user (
+	id varchar(50) PRIMARY KEY NOT NULL,
+	name varchar(255) NOT NULL,
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: username

File: pkg/db/schema/iam/V0_1__init.sql [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+	password varchar(255) NOT NULL,
+	email varchar(255) NOT NULL,
+	role varchar(32) NOT NULL,
+	description text
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: should add status fields
see repo

File: pkg/db/schema/iam/V0_1__init.sql [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS user_client (
+	user_id varchar(50) PRIMARY KEY NOT NULL,
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: not primary key

File: pkg/db/schema/iam/V0_1__init.sql [hunk 12]
```diff
@@ -0,0 +1,27 @@
+CREATE TABLE IF NOT EXISTS user (
+	id varchar(50) PRIMARY KEY NOT NULL,
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: s/id/user_id

File: pkg/db/schema/iam/V0_1__init.sql [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+CREATE TABLE IF NOT EXISTS user_client (
+	user_id varchar(50) PRIMARY KEY NOT NULL,
+	client_id varchar(50) PRIMARY KEY NOT NULL,
+	client_secret varchar(255) NOT NULL,
+	description text
+);
+
+CREATE TABLE IF NOT EXISTS group (
+	id varchar(50) PRIMARY KEY NOT NULL,
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: group_id

File: pkg/db/schema/iam/V0_1__init.sql [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+	name varchar(255) NOT NULL,
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: default is empty string

File: pkg/db/schema/iam/V0_1__init.sql [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+	description text
+);
+
+CREATE TABLE IF NOT EXISTS group_member (
+	group_id varchar(50) NOT NULL,
+	user_id varchar(50) NOT NULL,
+	UNIQUE INDEX group_member_uniq_index (group_id, user_id)
```
Comment (chai2010) [inline pkg/db/schema/iam/V0_1__init.sql]: group_id pri key
user_id pry key
delete UNIQUE

File: pkg/models/group_member.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+	"openpitrix.io/openpitrix/pkg/pb/iam"
+	"openpitrix.io/openpitrix/pkg/util/idutil"
+	"openpitrix.io/openpitrix/pkg/util/pbutil"
+)
+
+const GroupMemberTableName = "group_member"
+
+func NewGroupMemberId() string {
+	return idutil.GetUuid("grp-mem-")
+}
```
Comment (chai2010) [inline pkg/models/group_member.go]: delete

File: pkg/models/group_member.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+type GroupMember struct {
+	GroupId string
+	UserId  string
+}
+
+var GroupMemberColumns = GetColumnsFromStruct(&GroupMember{})
```
Comment (chai2010) [inline pkg/models/group_member.go]: delete

File: pkg/models/group_member.go [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+		UserId:  uid,
+	}
+}
+
+func GroupMemberToPb(p *GroupMember) *pbiam.GroupMember {
+	q := new(pbiam.GroupMember)
+	q.GroupId = pbutil.ToProtoString(p.GroupId)
+	q.UserId = pbutil.ToProtoString(p.UserId)
+	return q
+}
```
Comment (chai2010) [inline pkg/models/group_member.go]: delete

File: pkg/service/iam/handler_account.go [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+
+	reply := &pbiam.DeleteGroupsResponse{
+		GroupId: req.GroupId,
+	}
+
+	return reply, nil
+}
+
+func (p *Server) JoinGroup(ctx context.Context, req *pbiam.JoinGroupRequest) (*pbiam.JoinGroupResponse, error) {
+	if len(req.GroupId) != 1 || len(req.UserId) != 1 {
```
Comment (zheng1) [inline pkg/service/iam/handler_account.go]: Loop user_id and group_id to generate the record to be inserted. Check if the record exists before insertion and skip if it exists.
`len(req.GroupId) != 0 && len(req.UserId) != 0`

File: pkg/service/iam/handler_account.go [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+	reply := &pbiam.JoinGroupResponse{
+		GroupId: req.GroupId,
+		UserId:  req.UserId,
+	}
+
+	return reply, nil
+}
+
+func (p *Server) LeaveGroup(ctx context.Context, req *pbiam.LeaveGroupRequest) (*pbiam.LeaveGroupResponse, error) {
+	if len(req.GroupId) != 1 || len(req.UserId) != 1 {
```
Comment (zheng1) [inline pkg/service/iam/handler_account.go]:
```
len(req.GroupId) != 0 && len(req.UserId) != 0
```

File: pkg/service/iam/handler_account.go [hunk 21]
```diff
@@ -0,0 +1,442 @@
...
+		return nil, fmt.Errorf("missing GroupId or UserId")
+	}
+
+	var lastErr error
+	var reply = &pbiam.LeaveGroupResponse{}
+
+	for _, gid := range req.GroupId {
```
Comment (zheng1) [inline pkg/service/iam/handler_account.go]: There is no need for looping, the database will handle the IN operation

**Pull request closed without merging**